crates/ruff_db/src/vfs.rs

```diff
@@ -0,0 +1,318 @@
+                        FileRevision::zero(),
+                        FileStatus::Deleted,
+                        Count::default(),
+                    ),
+                }
+            })
+    }
+
+    /// Lookups a vendored file by its path. Returns `Some` if a vendored file for the given path
+    /// exists and `None` otherwise.
+    pub fn vendored(&self, db: &dyn Db, path: &VendoredPath) -> Option<VfsFile> {
+        let file = match self
+            .inner
+            .files_by_path
+            .entry(VfsPath::Vendored(path.to_path_buf()))
+        {
+            Entry::Occupied(entry) => *entry.get(),
+            Entry::Vacant(entry) => {
+                let revision = self.inner.vendored.revision(path)?;
+
+                let file = VfsFile::new(
+                    db,
+                    VfsPath::Vendored(path.to_path_buf()),
+                    Some(0o444),
+                    revision,
+                    FileStatus::Exists,
+                    Count::default(),
+                );
+
+                entry.insert(file);
+
+                file
+            }
+        };
+
+        Some(file)
+    }
+
+    /// Stubs out the vendored files with the given content.
+    ///
+    /// ## Panics
+    /// If there are pending snapshots referencing this `Vfs` instance.
+    pub fn stub_vendored<P, S>(&mut self, vendored: impl IntoIterator<Item = (P, S)>)
+    where
+        P: AsRef<VendoredPath>,
+        S: ToString,
+    {
+        let inner = Arc::get_mut(&mut self.inner).unwrap();
+
+        let stubbed = FxDashMap::default();
+
+        for (path, content) in vendored {
+            stubbed.insert(path.as_ref().to_path_buf(), content.to_string());
+        }
+
+        inner.vendored = VendoredVfs::Stubbed(stubbed);
+    }
+
+    /// Creates a salsa like snapshot of the files. The instances share
+    /// the same path to file mapping.
+    pub fn snapshot(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+
+    fn read(&self, db: &dyn Db, path: &VfsPath) -> String {
+        match path {
+            VfsPath::FileSystem(path) => db.file_system().read(path).unwrap_or_default(),
+
+            VfsPath::Vendored(vendored) => db
+                .vfs()
+                .inner
+                .vendored
+                .read(vendored)
+                .expect("Vendored file to exist"),
+        }
+    }
+}
+
+impl std::fmt::Debug for Vfs {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut map = f.debug_map();
+
+        for entry in self.inner.files_by_path.iter() {
+            map.entry(entry.key(), entry.value());
+        }
+        map.finish()
+    }
+}
+
+#[salsa::input]
+pub struct VfsFile {
+    /// The path of the file.
+    #[id]
+    #[return_ref]
+    pub path: VfsPath,
+
+    /// The unix permissions of the file. Only supported on unix systems. Always 0 on Windows
+    /// or when the file has been deleted.
```

Comment:
It should be `None`. 

The FS api uses a `u32` as return type. I'm not aware that 0 is ever a valid permission but I want to avoid conversion errors and using a `NonZeroU32` doesn't give us much here (other than the size savings)